pkgs/by-name/co/comfyui/package.nix

```diff
@@ -0,0 +1,151 @@
+{ lib
+, python311
+, linkFarm
+, writers
+, writeTextFile
+, fetchFromGitHub
+, stdenv
+, symlinkJoin
+, config
+, gpuBackend ? (
+  if config.cudaSupport
+  then "cuda"
+  else if config.rocmSupport
+  then "rocm"
+  else "none"
+)
+, modelsPath ? "/var/lib/comfyui/models"
+, inputPath ? "/var/lib/comfyui/input"
+, outputPath ? "/var/lib/comfyui/output"
+, tempPath ? "/var/lib/comfyui/temp"
+, userPath ? "/var/lib/comfyui/user"
+, customNodes ? []
+, models ? {
+    checkpoints = {};
+    clip = {};
+    clip_vision = {};
+    configs = {};
+    controlnet = {};
+    embeddings = {};
+    upscale_modules = {};
+    vae = {};
+    vae_approx = {};
+  }
+}:
+
+let
+
+  config-data = {
+    comfyui = {
+      base_path = modelsPath;
+      checkpoints = "${modelsPath}/checkpoints";
+      clip = "${modelsPath}/clip";
+      clip_vision = "${modelsPath}/clip_vision";
+      configs = "${modelsPath}/configs";
+      controlnet = "${modelsPath}/controlnet";
+      embeddings = "${modelsPath}/embeddings";
+      loras = "${modelsPath}/loras";
+      upscale_models= "${modelsPath}/upscale_models";
+      vae = "${modelsPath}/vae";
+      vae_approx = "${modelsPath}/vae_approx";
+    };
+  };
+
+  modelPathsFile = writeTextFile {
+    name = "extra_model_paths.yaml";
+    text = (lib.generators.toYAML {} config-data);
+  };
+
+  pythonEnv = (python311.withPackages (ps: with ps; [
+    pytorch-bin
+    # (
+    #   if gpuBackend == "cuda"
+    #   then torchWithCuda
+    #   else if gpuBackend == "rocm"
+    #   then torchWithRocm
+    #   else torch
+    # )
+    # torchsde
+    torchvision-bin
+    # torchaudio
+    (transformers.override { torch = pytorch-bin; })
+    (safetensors.override { torch = pytorch-bin; })
+    (accelerate.override { torch = pytorch-bin; })
```

Comment:
Regarding `pytorch-bin`: we should abstain from referring to `pytorch-bin` directly and from using ad hoc overrides like that. Instead, we should consistently refer to Nixpkgs' attributes, so that users can expect their overlays to actually have effect.

If a particular user decides to use pytorch-bin they'll just import nixpkgs with something like `{ overlays = [(final: prev: { pythonPackagesExtensions = [ (py-fin: py-pre: { torch = py-fin.pytorch-bin; }) ]; })];`

Reply:
Thanks!  Yes this is the guidance I was looking for because I knew it was wrong.  I'll give this a try and see if it still works for me.

Reply:
Hmm, this hasn't worked for me in the end.  I've spun my wheels quite a bit on getting this working.  Currently I'm stuck on this when using this with `pytorch-bin` / `torch-bin` as above:

```
       error: function 'anonymous lambda' called with unexpected argument 'cudaSupport'

       at /nix/store/c7xdr8fi9vbjmp2srk61cidhc9azb9yf-source/pkgs/development/python-modules/torch/bin.nix:1:1:

```

And building `torch` just remains out of reach for me.  The exact error seems to vary based on my exact commit of `nixpkgs` I'm using.  It _almost_ looks like a hardware error, but I've exercised the machine more consistently and extensively than by building `torch` (such as actually using it).  The build machine has 32 GB of ram and it's only used about 200 GB of its total 1 TB of disk.  Some advice suggests resource exhaustion.

I'm kind of stuck on this.  I'll keep trying things.

FWIW I periodically rebase atop `master` so at least I'm not chasing last month's fixed bugs.

Reply:
I was bit by a variant of https://github.com/NixOS/nixpkgs/issues/138943 (which I think is a separate but related issue from https://github.com/NixOS/nixpkgs/issues/255070).  I've worked past it now.  My latest force-push removes all of this `torch-bin` forcing since I have demonstrated everything works without it for myself.  Thanks!